homeassistant/components/mazda/button.py

```diff
@@ -0,0 +1,140 @@
+):
+    """Describes a Mazda button entity."""
+
+    # Function to determine whether the vehicle supports this button, given the coordinator data
+    is_supported: Callable[[dict[str, Any]], bool] = lambda data: True
+
+    async_press: Callable[[MazdaButtonEntity], Awaitable] = handle_button_press
+
+
+BUTTON_ENTITIES = [
+    MazdaButtonEntityDescription(
+        key="start_engine",
+        name_suffix="Start Engine",
+        icon="mdi:engine",
+    ),
+    MazdaButtonEntityDescription(
+        key="stop_engine",
+        name_suffix="Stop Engine",
+        icon="mdi:engine-off",
+    ),
+    MazdaButtonEntityDescription(
+        key="turn_on_hazard_lights",
+        name_suffix="Turn On Hazard Lights",
+        icon="mdi:hazard-lights",
+    ),
+    MazdaButtonEntityDescription(
+        key="turn_off_hazard_lights",
+        name_suffix="Turn Off Hazard Lights",
+        icon="mdi:hazard-lights",
+    ),
+    MazdaButtonEntityDescription(
+        key="start_charging",
+        name_suffix="Start Charging",
+        icon="mdi:ev-station",
+        is_supported=lambda data: data["isElectric"],
+    ),
+    MazdaButtonEntityDescription(
+        key="stop_charging",
+        name_suffix="Stop Charging",
+        icon="mdi:ev-station",
+        is_supported=lambda data: data["isElectric"],
+    ),
+    MazdaButtonEntityDescription(
+        key="refresh_vehicle_status",
+        name_suffix="Refresh Status",
+        icon="mdi:refresh",
+        async_press=handle_refresh_vehicle_status,
+        is_supported=lambda data: data["isElectric"],
+    ),
+]
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the button platform."""
+    client = hass.data[DOMAIN][config_entry.entry_id][DATA_CLIENT]
+    coordinator = hass.data[DOMAIN][config_entry.entry_id][DATA_COORDINATOR]
+
+    async_add_entities(
+        MazdaButtonEntity(client, coordinator, index, description)
+        for index, data in enumerate(coordinator.data)
+        for description in BUTTON_ENTITIES
+        if description.is_supported(data)
+    )
+
+
+class MazdaButtonEntity(MazdaEntity, ButtonEntity):
+    """Representation of a Mazda button."""
+
+    entity_description: MazdaButtonEntityDescription
+
+    def __init__(
+        self,
+        client: MazdaAPIClient,
+        coordinator: DataUpdateCoordinator,
+        index: int,
+        description: MazdaButtonEntityDescription,
+    ) -> None:
+        """Initialize Mazda button."""
+        super().__init__(client, coordinator, index)
+        self.entity_description = description
+
+        self._attr_name = f"{self.vehicle_name} {description.name_suffix}"
+        self._attr_unique_id = f"{self.vin}_{description.key}"
+
+    async def async_press(self) -> None:
+        """Press the button."""
+        await self.entity_description.async_press(self)
```

Comment:
Generally we don't want to pass around the entity.

Please pass the needed objects instead